tests/mock_crust/client_restrictions.rs

```diff
@@ -215,3 +219,132 @@ fn resend_over_load() {
 }
+
+// Sends the requests and verifies recipients if `rate_limiter` approves the request.
+// Returns rejected requests' message id and its client's ip address.
+fn rate_limiter_send_reqs(
+    network: &Network<PublicId>,
+    nodes: &mut [TestNode],
+    clients: &mut [TestClient],
+    total_usage: &mut u64,
+    is_always_send: bool,
+) -> HashMap<MessageId, IpAddr> {
+    let mut rng = network.new_rng();
+    let data_id: XorName = rng.gen();
+    let dst = Authority::NaeManager(data_id);
+
+    let mut clients_sent = HashMap::new();
+    for client in clients.iter_mut() {
+        if is_always_send || rng.gen_weighted_bool(2) {
+            let msg_id = MessageId::new();
+            unwrap!(client.inner.get_idata(dst, data_id, msg_id));
+            let _ = clients_sent.insert(msg_id, client.ip());
+        }
+    }
+    trace!("clients_sent: {:?}", clients_sent);
+    let _ = poll_all(nodes, clients);
+
+    let mut request_received: HashMap<MessageId, usize> = HashMap::new();
+    for node in nodes.iter_mut().filter(|n| n.is_recipient(&dst)) {
+        while let Ok(event) = node.try_next_ev() {
+            if let Event::Request {
+                request: Request::GetIData { msg_id: req_message_id, .. }, ..
+            } = event
+            {
+                let entry = request_received.entry(req_message_id).or_insert(0);
+                *entry += 1;
+            }
+        }
+    }
+    trace!("request_received: {:?}", request_received);
+
+    for (msg_id, count) in &request_received {
+        assert_eq!(*count, MIN_SECTION_SIZE);
+        let _ = clients_sent.remove(msg_id);
+        *total_usage += MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
+    }
+
+    // Max bandwidth used is C + C/2 + C/3 + C/4 + ... + C/2 = C * HarmonicNumber(N)
+    // FIXME(michael): this isn't really the max bandwidth, it can get higher somehow,
+    // but I haven't worked out how yet.
+    let cap = ::std::cmp::max(
+        SOFT_CAPACITY * harmonic_number(clients.len()).ceil() as u64,
+        MIN_CLIENT_CAPACITY * clients.len() as u64,
+    );
+    assert!(*total_usage <= cap, "{} > {}", total_usage, cap);
+
+    clients_sent
+}
+
+fn harmonic_number(n: usize) -> f64 {
+    (1..n + 1).map(|i| 1.0 / i as f64).sum()
+}
+
+// Verifies the usage in rate_limiter. Also advances the fake clock.
+fn rate_limiter_verify(
+    rejected_reqs: &HashMap<MessageId, IpAddr>,
+    clients_usage: &BTreeMap<IpAddr, u64>,
+    total_usage: &mut u64,
+    _per_client_cap: u64,
+) {
+    // `rejected_reqs` contains only the clients whose request got rejected.
+    // Needs to confirm such rejection is valid. However, if it is the total usage reaching the
+    // cap, the usage of each client could be much less than the cap.
+    if (*total_usage + MAX_IMMUTABLE_DATA_SIZE_IN_BYTES) <= RATE as u64 {
+        for ip in rejected_reqs.values() {
+            let client_usage = unwrap!(clients_usage.get(ip));
+            assert!(*client_usage >= MIN_CLIENT_CAPACITY);
```

Comment:
This whole verification step is trickier with the new rules. The concept of a "per client cap" is no longer really meaningful because the caps are constantly changing. This assert just says that if a client had a request rejected, it's because its usage was already above the min capacity, even if its actually capacity at the time the request was rejected may have been much larger.